docs/proposals/20231121-high-availability-of-edge-services.md

```diff
@@ -0,0 +1,282 @@
+
+### Goals
+
+- Design edge load balance controllers to monitor and distribute VIPs.
+- Enable VIP management from the level of the nodepool.
+- Design and implement VIP generation and service high availability based on Keepalived.
+- Design round rotation based ipvs module for traffic forwarding strategy.
+
+### Non-Goals/Future Work
+
+- While efficient load balancing is provided, users need to carefully plan the VIP for each nodepool. Especially in a LAN, if there are two nodepools, and both are bound to the same VIP, this could cause conflicts and errors. These issues arise from erroneous VIP planning, for which the LB is not responsible.
+- Provide more traffic forwarding policies, such as least connection, random, hash and so on.
+
+## Proposal
+
+### Architecture
+
+<div align="center">
+  <img src = "../img/YurtLB/arch.png" width="800">
+</div>
+
+### Implementation Details
+
+#### YurtLB Controller
+
+The YurtLB Controller monitors changes in the service and the nodepool, implementing service load balancing for the edge nodes via Virtual IPs (VIPs).
+
+YurtLB controller can be described as:
+
+##### Reconcile
+
+1. YurtLB Controller will list/monitor NodePool CRs. User-specified VIP information will be uniformly added to the properties of the NodePool, irrespective of how (and if) the IP is advertised.
+
+```go
+// NodePoolSpec defines the desired state of NodePool
+type NodePoolSpec struct {
+	//...
+  Addresses []string `json:"addresses"`
+}
+```
+
+Here is a example for Edge NodePool in load balance mode:
+
+```yaml
+apiVersion: apps.openyurt.io/v1beta1
+kind: NodePool
+metadata:
+  name: beijing
+spec:
+  type: Edge
+  addresses:
+    - 2002:2:2::1-2002:2:2::100
+    - 172.20.0.100-172.20.0.200
+```
+
+2. When a load balanced service is created, an event triggers the YurtLB controller to assign a VIP to the service from a pre-defined range for the nodepool.
+
+- YurtLB Controller selects the appropriate VIP for the load-balanced service using customized assignment policies (round robin, random, etc.). It's important that the VIP chosen from the nodepool's VIP set isn't selected repeatedly. In cases of repeated allocation, YurtLB controller will retain only the first successful allocation.
+- YurtLB Controller will deliver the selected VIP information to the nodes where the load balancer service is located, including the nodes where the pods reside.
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: nginx
+  namespace: default
```

Comment:
I think we need to give users the options of which nodepool LBs are created in, not all nodepools with associated pods. so it is a good idea to use nodepool selector in annotations for selecting the target nodepools as following:

```
Annotations:
    service.openyurt.io/pool-labelselector: key1=val1,key2=val2
```

all nodepools with lables: key1: val1 and key2:val2 will be selected, and LB instance will be created in these nodepools.

Reply:
@River-sh For the annotation definition, do you have any comments?

Reply:
I recommend adopting `service.openyurt.io/nodepool-labelselector: key1=val1,key2=val2`